### PR TITLE
Adds permissions into Che multiuser.

### PR DESCRIPTION
**Pull request by mshaposhnik**
Adds permissions framework into Che multiuser version.

### CODE REVIEW

File: assembly-multiuser/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MultiUserCheWsMasterModule.java [hunk 1]
```diff
@@ -11,14 +11,35 @@
...
     bind(DataSource.class).toProvider(org.eclipse.che.core.db.JndiDataSourceProvider.class);
+    bind(org.eclipse.che.api.system.server.SystemServicePermissionsFilter.class);
+    install(new SystemPermissionsJpaModule());
+    install(new org.eclipse.che.api.permission.server.PermissionsModule());
+    install(new org.eclipse.che.api.workspace.server.WorkspaceApiPermissionsModule());
+    bind(UserProfileServicePermissionsFilter.class);
+    bind(UserServicePermissionsFilter.class);
+    bind(FactoryPermissionsFilter.class);
+    bind(ActivityPermissionsFilter.class);
+    bindConstant().annotatedWith(Names.named("system.super_privileged_mode")).to(false);
```
Comment (sleshchenko) [inline assembly-multiuser/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MultiUserCheWsMasterModule.java]: Should not be it defined in property file to make admin able to override this property by external configuration?

File: wsmaster/che-core-api-authorization-impl/src/test/resources/logback-test.xml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+<configuration>
+
+    <appender name="stdout" class="ch.qos.logback.core.ConsoleAppender">
+        <encoder>
+            <pattern>%-41(%date[%.15thread]) %-45([%-5level] [%.30logger{30} %L]) - %msg%n</pattern>
+        </encoder>
+    </appender>
+
+    <appender name="file" class="ch.qos.logback.core.FileAppender">
+        <File>target/log/codenvy.log</File>
```
Comment (sleshchenko) [inline wsmaster/che-core-api-authorization-impl/src/test/resources/logback-test.xml]: It's better to rename this log file if we need it.

File: wsmaster/che-core-api-machine/src/test/java/org/eclipse/che/api/machine/server/jpa/RecipePermissionsTckModule.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import org.eclipse.che.core.db.schema.SchemaInitializer;
+import org.eclipse.che.core.db.schema.impl.flyway.FlywaySchemaInitializer;
+
+/** @author Max Shaposhnik */
+public class RecipePermissionsTckModule extends TckModule {
+  @Override
+  protected void configure() {
+    install(new JpaPersistModule("main"));
+    bind(SchemaInitializer.class)
+        .toInstance(new FlywaySchemaInitializer(inMemoryDefault(), "che-schema", "codenvy-schema"));
```
Comment (sleshchenko) [inline wsmaster/che-core-api-machine/src/test/java/org/eclipse/che/api/machine/server/jpa/RecipePermissionsTckModule.java]: Does `che` really has `codenvy-schema`?

File: wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/jpa/JpaRecipeDao.java [hunk 4]
```diff
@@ -101,33 +104,60 @@ public RecipeImpl getById(String id) throws NotFoundException, ServerException {
     }
   }
 
+  /**
+   * Translated query should look like:
+   *
+   * <p>SELECT recipe.ID, recipe.CREATOR, recipe.DESCRIPTION, recipe.NAME, recipe.SCRIPT,
```
Comment (sleshchenko) [inline wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/jpa/JpaRecipeDao.java]: IMHO it is really difficult to read this query. Please reformat it

File: wsmaster/che-core-api-machine/src/test/resources/META-INF/persistence.xml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    Contributors:
+      Red Hat, Inc. - initial API and implementation
+
+-->
+<persistence xmlns="http://java.sun.com/xml/ns/persistence"
+             xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+             xsi:schemaLocation="http://java.sun.com/xml/ns/persistence persistence_1_0.xsd" version="1.0">
+    <persistence-unit name="main" transaction-type="RESOURCE_LOCAL">
+        <provider>org.eclipse.persistence.jpa.PersistenceProvider</provider>
```
Comment (sleshchenko) [inline wsmaster/che-core-api-machine/src/test/resources/META-INF/persistence.xml]: Would not be better to use PersistTestModuleBuilder here?

File: wsmaster/che-core-api-authorization-impl/src/test/resources/logback-test.xml (hunk 2 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-authorization-impl/src/test/resources/logback-test.xml]: Ok

File: wsmaster/che-core-api-machine/src/test/java/org/eclipse/che/api/machine/server/jpa/RecipePermissionsTckModule.java (hunk 3 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-machine/src/test/java/org/eclipse/che/api/machine/server/jpa/RecipePermissionsTckModule.java]: C/P error

File: assembly-multiuser/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MultiUserCheWsMasterModule.java [hunk 6]
```diff
@@ -11,14 +11,35 @@
...
 
 @DynaModule
 public class MultiUserCheWsMasterModule extends AbstractModule {
 
   @Override
   protected void configure() {
     bind(DataSource.class).toProvider(org.eclipse.che.core.db.JndiDataSourceProvider.class);
+    bind(org.eclipse.che.api.system.server.SystemServicePermissionsFilter.class);
+    install(new SystemPermissionsJpaModule());
+    install(new org.eclipse.che.api.permission.server.PermissionsModule());
```
Comment (sleshchenko) [inline assembly-multiuser/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MultiUserCheWsMasterModule.java]: let's bind classes or install modules with fqn or without, but in the same way

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/5.17.0/1__add_permissions.sql [hunk 7]
```diff
@@ -0,0 +1,126 @@
+--
```
Comment (sleshchenko) [inline wsmaster/che-core-sql-schema/src/main/resources/che-schema/5.17.0/1__add_permissions.sql]: Looks like you should move this script to 5.18.0 folder

File: dockerfiles/init/manifests/che.env [hunk 8]
```diff
@@ -429,6 +429,8 @@ CHE_SINGLE_PORT=false
 #     property 'che.workspace.auto_snapshot=false'.
 #CHE_WORKSPACE_AUTO__SNAPSHOT=false
 
+#SYSTEM_SUPER_PRIVILEGED_MODE=false
```
Comment (sleshchenko) [inline dockerfiles/init/manifests/che.env]: I'd add a description for this property. See https://github.com/codenvy/codenvy/blob/master/dockerfiles/init/manifests/codenvy.env#L40

File: wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/jpa/JpaRecipeDao.java (hunk 4 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/jpa/JpaRecipeDao.java]: I can not due to fmt plugin :(

Comment (sleshchenko) [reply wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/jpa/JpaRecipeDao.java]: Formatting in this way works fine for me
```
/**
   * Translated query should look like:
   *
   * <pre>
   * SELECT recipe.ID, recipe.CREATOR, recipe.DESCRIPTION, 
   *     recipe.NAME, recipe.SCRIPT, recipe.TYPE
   * FROM {oj RECIPEPERMISSIONS permission
   * LEFT OUTER JOIN RECIPE recipe ON (recipe.ID = permission.RECIPEID)
   * LEFT OUTER JOIN Recipe_TAGS tag ON (tag.Recipe_ID = recipe.ID)}, 
   *     RECIPEPERMISSIONS_ACTIONS permissionActions
   * WHERE ((tag.tag IN (?)) 
   *     AND ((? IS NULL) OR (recipe.TYPE IS NULL) OR (recipe.TYPE = ?)) 
   *     AND ((permission.USERID IS NULL) OR (permission.USERID = ?)) 
   *     AND (permissionActions.actions = ?) 
   *     AND (permissionActions.RECIPEPERMISSIONS_ID = permission.ID))
   * GROUP BY recipe.ID 
   * HAVING (COUNT(tag.tag) = ?)
   * </pre>
   */
```

File: dockerfiles/init/manifests/che.env [hunk 9]
```diff
@@ -429,7 +429,13 @@ CHE_SINGLE_PORT=false
 #     property 'che.workspace.auto_snapshot=false'.
 #CHE_WORKSPACE_AUTO__SNAPSHOT=false
 
-#SYSTEM_SUPER_PRIVILEGED_MODE=false
+# System Super Privileged Mode
+#     Grants users with the manageSystem permission additional permissions for
+#     getByKey, getByNameSpace, stopWorkspaces, and getResourcesInformation.
```
Comment (sleshchenko) [inline dockerfiles/init/manifests/che.env]: It'd be better to remove information about ResourcesService while it is not moved to che.

Comment (mshaposhnik) [reply dockerfiles/init/manifests/che.env]: it will be moved soon as well

Comment (sleshchenko) [reply dockerfiles/init/manifests/che.env]: Thanks. Then OK to leave it

File: wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/jpa/JpaRecipeDao.java [hunk 10]
```diff
@@ -101,33 +104,74 @@ public RecipeImpl getById(String id) throws NotFoundException, ServerException {
...
+   * Translated query should look like:
+   *
+   * <pre>
+   * SELECT recipe.ID,
+   *        recipe.CREATOR,
+   *        recipe.DESCRIPTION,
+   *        recipe.NAME,
+   *        recipe.SCRIPT,
+   *        recipe.TYPE
+   * FROM {oj RECIPEPERMISSIONS permission
```
Comment (sleshchenko) [inline wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/jpa/JpaRecipeDao.java]: IWhat does `{oj` mean here?

Comment (mshaposhnik) [reply wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/jpa/JpaRecipeDao.java]: Don;t know. Just copied from [original](https://github.com/codenvy/codenvy/blob/master/wsmaster/codenvy-hosted-api-machine/src/main/java/com/codenvy/api/machine/server/jpa/OnPremisesJpaRecipeDao.java#L49)  Probably shortcut of "OUTER JOIN"

Comment (akorneta) [reply wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/jpa/JpaRecipeDao.java]: @mshaposhnik you can remove this. It's was added only 
for clarifying of the resulting query.

**Review by skabashnyuk — APPROVED** [review]

File: wsmaster/che-core-api-permission/src/test/resources/META-INF/persistence.xml [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+-->
+<persistence xmlns="http://java.sun.com/xml/ns/persistence"
+             xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+             xsi:schemaLocation="http://java.sun.com/xml/ns/persistence persistence_1_0.xsd" version="1.0">
+    <persistence-unit name="main" transaction-type="RESOURCE_LOCAL">
+        <provider>org.eclipse.persistence.jpa.PersistenceProvider</provider>
+        <class>org.eclipse.che.api.permission.server.model.impl.AbstractPermissions</class>
+        <class>org.eclipse.che.api.permission.server.model.impl.SystemPermissionsImpl</class>
+
+        <class>org.eclipse.che.account.spi.AccountImpl</class>
```
Comment (sleshchenko) [inline wsmaster/che-core-api-permission/src/test/resources/META-INF/persistence.xml]: Let's use `PersistTestModuleBuilder` here

**Review by sleshchenko — APPROVED** [review]
Other LGTM

**Pull request merged**